`Web docs` - Key value inputs (#2941)
hashicorp/design-system: website/docs/components/form/key-value-inputs/partials/code/how-to-use.md

## How to use this component

The basic invocation of the Key Value Inputs requires the `@data` argument. This is used as the initial data to create the rows of inputs.

```handlebars
<Hds::Form::KeyValueInputs @data={{array
    (hash name="Judith Maxene" email="j.maxene@randatmail.com")
    (hash name="Elmira Aishah" email="e.aishah@randatmail.com")
  }}>
  <:header as |H|>
    <H.Legend>Invite multiple users to your organization</H.Legend>
    <H.HelperText>
      Users without organization roles cannot view or edit anything inside their organization until project-level or workspace-level roles are assigned to them after they accept their invitation.
    </H.HelperText>
  </:header>
  <:row as |R|>
    <R.Field as |F|>
      <F.Label>
        Email address for invitee
      </F.Label>
      <F.TextInput @value={{R.rowData.email}} />
    </R.Field>
    <R.Field as |F|>
      <F.Label>
        Name of invitee
      </F.Label>
      <F.TextInput @value={{R.rowData.name}} />
    </R.Field>
    <R.DeleteRowButton />
  </:row>
  <:footer as |F|>
    <F.AddRowButton @text="Add user" />
  </:footer>
</Hds::Form::KeyValueInputs>
```

### Required vs. optional

> Use the `@isRequired` and `@isOptional` arguments to add a visual indicator that the entire Key Value Inputs is required or optional.

Reviewer: Maybe `KeyValuePair` would have worked a little better.

    (hash name="Judith Maxene" email="j.maxene@randatmail.com")
  }}>
  <:header as |H|>
    <H.Legend>Invite multiple users to your organization</H.Legend>
  </:header>
  <:row as |R|>
    <R.Field as |F|>
      <F.Label>
        Email address for invitee
      </F.Label>
      <F.TextInput @value={{R.rowData.email}} />
    </R.Field>
    <R.Field as |F|>
      <F.Label>
        Name of invitee
      </F.Label>
      <F.TextInput @value={{R.rowData.name}} />
    </R.Field>
    <R.DeleteRowButton />
  </:row>
</Hds::Form::KeyValueInputs>
<hr />
<Hds::Form::KeyValueInputs @isOptional={{true}} @data={{array
    (hash name="Judith Maxene" email="j.maxene@randatmail.com")
  }}>
  <:header as |H|>
    <H.Legend>Invite multiple users to your organization</H.Legend>
  </:header>
  <:row as |R|>
    <R.Field as |F|>
      <F.Label>
        Email address for invitee
      </F.Label>
      <F.TextInput @value={{R.rowData.email}} />
    </R.Field>
    <R.Field as |F|>
      <F.Label>
        Name of invitee
      </F.Label>
      <F.TextInput @value={{R.rowData.name}} />
    </R.Field>
    <R.DeleteRowButton />
  </:row>
</Hds::Form::KeyValueInputs>
```
### Validation

> The Key Value Inputs can be validated at the fieldset or each individual field. You can choose which level is appropriate depending on the use case, but it is best practice to make any error message as specific as possible. Avoid putting the error on the fieldset if the issue is with a single field.

Reviewer 2: ```suggestion
The Key Value Inputs can be validated at the fieldset level or on each individual field. You can choose which level is appropriate depending on the use case, but it is best practice to make any error message as specific as possible. Avoid putting the error on the fieldset if the issue occurs on a single field.
```

> The Key Value Inputs can be validated at the fieldset or each individual field. You can choose which level is appropriate depending on the use case, but it is best practice to make any error message as specific as possible. Avoid putting the error on the fieldset if the issue is with a single field.

Reviewer 2: ```suggestion
The Key Value Inputs can be validated at the fieldset level or on each individual field. You can choose which level is appropriate depending on the use case, but it is best practice to make any error message as specific as possible. Avoid putting the error on the fieldset if the issue occurs on a single field.
```

    <H.Legend>Invite multiple users to your organization</H.Legend>
    <H.HelperText>
      Users without organization roles cannot view or edit anything inside their organization until project-level or workspace-level roles are assigned to them after they accept their invitation.
    </H.HelperText>
  </:header>
  <:row as |R|>
    <R.Field @isInvalid={{true}} as |F|>
      <F.Label>
        Email address for invitee
      </F.Label>
      <F.TextInput @value={{R.rowData.email}} />
      <F.Error>Email is required.</F.Error>
    </R.Field>
    <R.Field as |F|>
      <F.Label>
        Name of invitee
      </F.Label>
      <F.TextInput @value={{R.rowData.name}} />
    </R.Field>
    <R.DeleteRowButton />
  </:row>
  <:footer as |F|>
    <F.AddRowButton @text="Add user" />
  </:footer>
</Hds::Form::KeyValueInputs>
```

To display an error for the whole group of inputs, use the `[F].Error` component yielded by the `:footer` named block.

```handlebars
<Hds::Form::KeyValueInputs @data={{array
    (hash name="Judith Maxene" email="j.maxene@randatmail.com")
  }}>
  <:header as |H|>
    <H.Legend>Invite multiple users to your organization</H.Legend>
    <H.HelperText>

>       Users without organization roles cannot view or edit anything inside their organization until project-level or workspace-level roles are assigned to them after they accept their invitation.

Reviewer: ```suggestion
      Users without an organization role cannot view or edit anything inside their organization until a project-level or workspace-level role is assigned to them after they accept their invitation.
```

Reviewer: Edited the above on the assumption that a user only need one appropriate-level role rather than needing to have multiple roles assigned to them.

```handlebars
<Hds::Form::KeyValueInputs @data={{array
    (hash os="darwin")
    (hash os="linux")
    (hash os="windows")
  }}>
  <:header as |H|>
    <H.Legend>Plugin binaries</H.Legend>
    <H.HelperText>
      A zipped binary file for each supported OS and architecture pair.
    </H.HelperText>
  </:header>
  <:row as |R|>
    <R.Field as |F|>
      <F.Label>
        OS and architecture
      </F.Label>
      <F.Select>
        <option selected={{if (eq R.rowData.os "darwin") true}}>darwin - arm64</option>
        <option selected={{if (eq R.rowData.os "linux") true}}>linux - 386</option>
        <option selected={{if (eq R.rowData.os "windows") true}}>windows - amd64</option>
      </F.Select>
    </R.Field>
    <R.Field as |F|>
      <F.Label>
       File
      </F.Label>
      <F.FileInput />
    </R.Field>
    <R.DeleteRowButton />
  </:row>
  <:footer as |F|>
    <F.AddRowButton @text="Add configuration" />
  </:footer>
</Hds::Form::KeyValueInputs>
```

### Generic content 

> If there is additional content needed in the header or a need for a not supported input type, there are generic slots available.

Reviewer: Perhaps you could add an example using simple placeholders for content to make the structure & usage more clear?

**Suggested edit:**

- "Generic slots are available to add additional content or unsupported input types."

> If there is additional content needed in the header or a need for a not supported input type, there are generic slots available.

Reviewer: ```suggestion
Generic slots are available to add additional content or unsupported input types.
```

```handlebars
<Hds::Form::KeyValueInputs @data={{array
    (hash email="j.maxene@randatmail.com" hasOrgRole=true)
    (hash email="e.aishah@randatmail.com" hasOrgRole=false)
  }}>
  <:header as |H|>
    <H.Legend>Grant access to HCP resources for your team members</H.Legend>
    <H.HelperText>

>       Users without organization roles cannot view or edit anything inside their organization until project-level or workspace-level roles are assigned to them after they accept their invitation.

Reviewer: ```suggestion
      Users without an organization role cannot view or edit anything inside their organization until a project-level or workspace-level role is assigned to them after they accept their invitation.
```

    </H.HelperText>
    <H.Generic>
      <Hds::Button @color="tertiary" @text="What can each role do?" @icon="help" />
    </H.Generic>
  </:header>
  <:row as |R|>
    <R.Field as |F|>
      <F.Label>
        Email address for invitee
      </F.Label>
      <F.TextInput @value={{R.rowData.email}} />
    </R.Field>
    <R.Generic>
      <Hds::Form::Toggle::Field checked={{R.rowData.hasOrgRole}} as |T|>
        <T.Label>Assign this user an organization role</T.Label>
      </Hds::Form::Toggle::Field>
    </R.Generic>
    <R.DeleteRowButton />
  </:row>
  <:footer as |F|>
    <F.AddRowButton @text="Add user" />
  </:footer>
</Hds::Form::KeyValueInputs>
```

### Updating the rows

> You are responsible for handling the logic to add or remove a row. To do this, use the `onClick` argument to pass a callback function on the yielded `[F].AddRowButton` in the `:footer` named block or the yielded `[R].DeleteRowButton` in the `:row` named block.

Reviewer: +1 That jumped out to me as well.

> You are responsible for handling the logic to add or remove a row. To do this, use the `onClick` argument to pass a callback function on the yielded `[F].AddRowButton` in the `:footer` named block or the yielded `[R].DeleteRowButton` in the `:row` named block.

Reviewer: ```suggestion
Consumers are responsible for handling the logic to add or remove a row. To do this, use the `onClick` argument to pass a callback function on the yielded `[F].AddRowButton` in the `:footer` named block or the yielded `[R].DeleteRowButton` in the `:row` named block.
```

```handlebars
<Hds::Form::KeyValueInputs @data={{array
    (hash os="darwin")
    (hash os="linux")
    (hash os="windows")
  }}>
  <:header as |H|>
    <H.Legend>Plugin binaries</H.Legend>
    <H.HelperText>
      A zipped binary file for each supported OS and architecture pair.
    </H.HelperText>
  </:header>
  <:row as |R|>
    <R.Field as |F|>
      <F.Label>
        OS and architecture
      </F.Label>
      <F.Select>
        <option selected={{if (eq R.rowData.os "darwin") true}}>darwin - arm64</option>
        <option selected={{if (eq R.rowData.os "linux") true}}>linux - 386</option>
        <option selected={{if (eq R.rowData.os "windows") true}}>windows - amd64</option>
      </F.Select>
    </R.Field>
    <R.Field as |F|>
      <F.Label>
       File
      </F.Label>
      <F.FileInput />
    </R.Field>
    <R.DeleteRowButton />
  </:row>
  <:footer as |F|>
    <F.Alert as |A|>
      <A.Description>Only 3 binaries can be added at a time.</A.Description>
    </F.Alert>
  </:footer>
</Hds::Form::KeyValueInputs>
```

### Field width

> By default, each field shares the available space evenly. There may be situations where you need to customize the width by setting the `width` argument on the field.

Reviewer: ```suggestion
Fields are equal width by default. Use the Field’s `width` argument to customize field widths.
```

## How to use this component

!!! Insight

> **TODO**

Reviewer 2: I was meant to remove it. Clearly I didn't 🤷‍♂️

> **TODO**
> 

Reviewer 2: ```suggestion
```

Given the complexity and level of custom code required to use this component, it is not possible to cover all the possible use cases in this documentation.

> For this reason, if you need to implement a key-value input pattern using this component, or migrate an existing one, [contact the Design Systems Team](/about/support) for support.

Reviewer 2: Yeah, I like it!

```handlebars
<!--  Note: this is a non-interactive example -->

<Hds::Form::KeyValueInputs @data={{array
    (hash email="j.maxene@randatmail.com" hasOrgRole=true)
    (hash email="e.aishah@randatmail.com" hasOrgRole=false)
  }}>
  <:header as |H|>
    <H.Legend>Grant access to HCP resources for your team members</H.Legend>
    <H.HelperText>
      Users without an organization role cannot view or edit anything inside their organization until a project-level or workspace-level role is assigned to them after they accept their invitation.
    </H.HelperText>
    <H.Generic>
      <Hds::Button @color="tertiary" @text="What can each role do?" @icon="help" />
    </H.Generic>
  </:header>
  <:row as |R|>
    <R.Field as |F|>
      <F.Label>
        Email address for invitee
      </F.Label>
      <F.TextInput @value={{R.rowData.email}} />
    </R.Field>
    <R.Generic>
      <Hds::Form::Toggle::Field checked={{R.rowData.hasOrgRole}} as |T|>
        <T.Label>Assign this user an organization role</T.Label>
      </Hds::Form::Toggle::Field>
    </R.Generic>
    <R.DeleteRowButton />
  </:row>
  <:footer as |F|>

>     <F.AddRowButton @text="Add user" />

Reviewer 2: As mentioned above, there's no yielded `Generic` block in the Footer
